### ISSUE TITLE
PHP 8.1 support

### ISSUE THREAD

**Issue by IonBazan** [aws/elastic-beanstalk-roadmap#214]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave "+1" or "me too" comments. They generate extra noise for issue followers and do not help prioritize the request.

**Tell us about your request**
What do you want us to build?

PHP 8.1 has officially been released on 25 November 2021, yay! Here are the changes and features introduced in this release: https://www.php.net/releases/8.1/en.php. It would be great to see it supported natively by Elastic Beanstalk.

**Are you currently working around this issue?**

Staying on PHP 8.0 or installing the 8.1 manually during startup.

**Comment by faizanakram99** [comment]
I had suggested earlier too that AWS should start working on new PHP version support while it is in beta state, now again we have a new stable PHP version (released one month ago) and there is no news to be seen about PHP 8.1 support.

https://github.com/aws/elastic-beanstalk-roadmap/issues/131#issuecomment-749805846

**Comment by misyaath** [comment]
And also in code build i could not build Laravel 9 because its needed  8.0.2 minimum version  Please update the version  and also i cannot update the version of php in code build environment when i install it dose not update php version

**Comment by kloining** [comment]
@misyaath Take a look at [PHP platform history](https://docs.aws.amazon.com/elasticbeanstalk/latest/platforms/platform-history-php.html)
By using the latest platform, Elastic Beanstalk runs on 8.0.13

**Comment by nickdnk** [comment]
Guys, come on. Assigning **one** person to making sure you have the most recent version of PHP available to your customers can't possibly be this hard.

With all due respect, why does this take so long? You've also sometimes been up to **6 months** behind on security updates for PHP versions you claim to support (7.4 and 8.0 for instance). That's borderline negligent.

**Comment by ghost** [comment]
This is so careless of aws beanstalk team. Such a shame

**Comment by Ciaro** [comment]
[nickdnk](https://github.com/nickdnk) pretty much nailed it with his comment...

**Comment by nickdnk** [comment]
What's the argument here, @devfrey and @axlon? Do you think this is all good or what? Why the downvotes?

**Comment by axlon** [comment]
@nickdnk Because these last few comments add nothing to this thread, regardless of if I agree with them or not. I am subscribed to this thread so that I'd get notified when they finally do decide to address this, but instead every time a notification does pop up its just more of the same.

Regardless, no hard feelings, I think we all want the same thing here.

**Comment by nickdnk** [comment]
@axlon That's fair, but you must also recognize that more complaints about this (and upvotes on the original post don't really generate any noise) should lead to a shift in priority within AWS, despite the fact that they claim it does not. It's not even **on** the roadmap yet or being responded to, even after 4 months and 120+ upvotes, which is ridiculous, to be quite honest.

**Comment by zainspresso** [comment]
For the time being I just do the update myself:
It is ridiculous how long this takes..

```
commands:
  00_upgrade_php:
    command: |
      sudo yum -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
      sudo yum -y install https://rpms.remirepo.net/enterprise/remi-release-7.rpm
      sudo yum -y install yum-utils
      sudo yum -y install rpl
      export LC_ALL=en_US.UTF-8
      export LANG=en_US.UTF-8
      sudo yum-config-manager --disable 'remi-php*'
      sudo yum-config-manager --enable remi-php81
      rpl [remi-php81] "[remi-php81]\npriority=1" /etc/yum.repos.d/remi-php81.repo -e
      yum upgrade -y php php-*
    test: test ! -f .semaphore
  01_lock:
    command: touch .semaphore
```

**Comment by wattsie** [comment]
Hi all,
  I was able to get the code build process working using similar by changing the following...

  You simply need to remove or disable phpenv, then install a newer php with curl and xml then all proceeds as expected.

  PS: you maybe able to just upgrade to 8.0.13, I just went straight to 8.1.

  PPS: also need to change "SolutionStackName" to "64bit Amazon Linux 2 v3.3.11 running PHP 8.0"

Cheers.

buildspec.yml
```
version: 0.2

phases:
  install:
    runtime-versions:
      php: 8.0
    commands:
      - rm -rf $(phpenv root)
      - add-apt-repository ppa:ondrej/php
      - apt-get update
      - apt-get install php8.1 php8.1-xml php8.1-curl -y
      - composer update
  build:
    commands:
      - rm -rf /var/app/current/public/storage
      - php artisan storage:link --force
  post_build:
    commands:
      - aws cloudformation package --template template.yml --s3-bucket $S3_BUCKET --output-template-file template-export.yml

      # Do not remove this statement. This command is required for AWS CodeStar projects.
      # Update the AWS Partition, AWS Region, account ID and project ID in the project ARN on template-configuration.json file so AWS CloudFormation can tag project resources.
      - sed -i.bak 's/\$PARTITION\$/'${PARTITION}'/g;s/\$AWS_REGION\$/'${AWS_REGION}'/g;s/\$ACCOUNT_ID\$/'${ACCOUNT_ID}'/g;s/\$PROJECT_ID\$/'${PROJECT_ID}'/g' template-configuration.json
artifacts:
  files:
    - template-export.yml
    - template-configuration.json
```

**Comment by vantezzen** [comment]
I've had some problems when using zainspresso's script while also using ImageMagick as there were some dependency conflicts during the upgrade.

I'm now using this under "/.ebextensions/00_upgrade_php.config" if anyone has the same problem:

```yml
commands:
  00_upgrade_php:
    command: |
      if [ `php -r "echo version_compare(PHP_VERSION, '8.1.0') >= 0;"` = 1 ]
      then
        echo "PHP already up-to-date"
      else
        sudo yum -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
        sudo yum -y install https://rpms.remirepo.net/enterprise/remi-release-7.rpm
        sudo yum -y install yum-utils
        sudo yum -y install rpl
        sudo yum-config-manager --setopt=remi.priority=1 --save
        # Reinstall ImageMagick from remi to prevent dependency conflicts
        sudo yum -y remove ImageMagick
        sudo yum -y install ImageMagick
        export LC_ALL=en_US.UTF-8
        export LANG=en_US.UTF-8
        sudo yum-config-manager --disable 'remi-php*'
        sudo yum-config-manager --enable remi-php81
        rpl [remi-php81] "[remi-php81]\npriority=1" /etc/yum.repos.d/remi-php81.repo -e
        yum upgrade -y php php-*
      fi
    test: test ! -f .semaphore
  01_lock:
    command: touch .semaphore
```

**Comment by juizmill** [comment]
It should have the latest version no later than a week after the official release.

**Comment by blaater6** [comment]
Almost half a year after the release of php 8.1 I'm really eager to use this on AWS. Maybe the maintainers can prioritise this issue so maybe we will have php 8.1 support when php 8.2 comes out.

**Comment by netadapt** [comment]
> Almost half a year after the release of php 8.1 I'm really eager to use this on AWS. Maybe the maintainers can prioritise this issue so maybe we will have php 8.1 support when php 8.2 comes out.

8.0 was available on Extras so promptly it set the expectation for quick availability of 8.1 too.  So this delay is baffling - I'm just pleased to have some update at all after all this time.   Those enums hopefully won't stay out of reach for much longer...

**Comment by TZwakhals** [comment]
@blaater6 Indeed, our project is kinda held back by the need for this update

**Comment by jordanade** [comment]
I want to reiterate that it's not just the major/minor upgrades that are delayed. The patch versions aren't getting updated either. Still on PHP 8.0.16 when 8.0.19 is out. There's a bug that's affecting us that's fixed in 8.0.17!

**Comment by RahulDey12** [comment]
Is there any update? what is happening AWS all other Linux distros already gave the update for PHP 8.1 but you guys are in AWS getting money from our bills and should have some responsibilities for their customers for providing the best support. When you guys are trying to promote your Amazon Linux 2, then why delay on PHP update when 7 out of 10 websites are made with PHP.

**Comment by KazakovVS** [comment]
Why does it take so long to support PHP 8.1?

**Comment by VaguelyOnline** [comment]
What is the hold up here?

**Comment by jdempster** [comment]
Also waiting. Would be good to get an update on when it’s happening

**Comment by winkelco** [comment]
how many years of work does aws need for this?

**Comment by KZeni** [comment]
Just a heads up, PHP 8.2 Alpha is out for testing as of today and yet this is still waiting to have 8.1 made available (with 8.1.7 being the current version now.)

Meanwhile, it's taken long enough that the bigger thing at this point is how https://www.php.net/supported-versions is showing version 8.0.x is losing active support in 5 months (we're past the halfway point from the official release and it losing active support) while this isn't offering any newer version than that to use/migrate to in order to stay with a version that has active support. Sure, it will have security support for a year after that, but this really should be offering what's recommended (has active support) and allow migration time for existing things and/or allow for the newest version to be used for new things. At the very least, 8.1 really should be made available sooner than later so those looking to use an actively supported PHP version have time to migrate, test, etc. before then.

**Comment by jordanade** [comment]
@KZeni What good is "security support" when we don't even get the patch upgrades?? Elastic beanstalk is a joke, I guess we should all be transitioning to Docker...which would also make it easier to dump AWS entirely and move to a different cloud provider.

**Comment by KZeni** [comment]
@jordanade I mean, https://docs.aws.amazon.com/elasticbeanstalk/latest/platforms/platform-history-php.html does show the PHP 8.0.x version has been getting updated over the past months (8.0.8 in July 2021, 8.0.13 in February 2022, 8.0.16 in April 2022, etc.) Those versions look to lag a few months behind their release when it comes to them being made available, but at least they're getting updated. It then just seems odd that 8.1.x is entirely missing at this point & isn't just a few months behind.

**Comment by sebastianpuka310** [comment]
We are also waiting for Elastic Beanstalk support for PHP 8.1. This is affecting our applications and holding them back. It would be good if we have an ETA when that would be available ? We hope it is not going to take long.

**Comment by ishu202** [comment]
same here i was going to deploy symphony 6.1 but had to rollback to symphony 6.0.9. not a big thing but missing out alot of optimizations and bug fixes.

**Comment by adamwinn** [comment]
The irony here is that if you add `"aws/aws-sdk-php": "3.*"` to your composer file, you will get an error 

```
Fatal error: Composer detected issues in your platform: Your Composer dependencies require a PHP version ">= 8.1.0". You are running 8.0.20. in /Users/me/vendor/composer/platform_check.php on line 24
```

So now you can't even use AWS's own PHP SDK with your PHP app because beanstalk hasn't been updated to support PHP 8.1

**Comment by girayk** [comment]
PHP 8.2 Alpha is out, almost 8.2 is coming and we are still waiting for 8.1

**Comment by Nilanth** [comment]
We are waiting for this release to update some predominant dependencies in our app. Kindly update the **ETA**.

**Comment by ishu202** [comment]
here is the solution i ended up with. i containerized my app using docker. and was still able to use RDS and other services. I could share my config if anyone is interested. let me know.

**Comment by faizanakram99** [comment]
> here is the solution i ended up with. i containerized my app using docker. and was still able to use RDS and other services. I could share my config if anyone is interested. let me know.

sure, go ahead

**Comment by jasonpeltzer** [comment]
Any status on this would be greatly appreciated, this is holding up our ability to move to Symfony 6.1!  Not to mention the lack of enums.  Oh yeah, and 8.2 is set for release!

**Comment by faizanakram99** [comment]
@devfrey you seem to have a lot of time to downvote each comment.

**Comment by andyexeter** [comment]
Happy to see that this has been moved to Coming Soon on the roadmap!

@devfrey please downvote this comment so I don't feel left out :)

**Comment by jordanade** [comment]
> Happy to see that this has been moved to Coming Soon on the roadmap!

Yeah, now we just have to "think a couple of months out, give or take"...

**Comment by nickdnk** [comment]
I have created an issue for PHP 8.2. We might as well get the ball rolling on that (given that its release date is closer than the current delay on 8.1) so we can avoid this type of situation going forward. Go and upvote it to support it.

Here: https://github.com/aws/elastic-beanstalk-roadmap/issues/250

**Comment by MarcoDeJong** [comment]
![Alt Text](https://media.giphy.com/media/26n6xBpxNXExDfuKc/giphy.gif)

**Comment by g3luka** [comment]
Hi guys,

I don't know if most of you use AWS services other than Beanstalk?
I also use AWS Lambda and platform updates are released the same way. So, it's not exclusive to either Beanstalk or PHP, with Node.js the same thing happens.

**Comment by CesarOsorioL** [comment]
Hi guys

Could you give us a date or something to know when **PHP 8.1** in Elastic Beanstalk will be available? This would be of great help, thank you.

**Comment by dejagersh** [comment]
> Hi guys
> 
> Could you give us a date or something to know when **PHP 8.1** in Elastic Beanstalk will be available? This would be of great help, thank you.

It's currently in status "Coming soon" since 23th of June.

PHP 8.0 moved into "Coming soon" on 27 feb 2021 and was released on 31 march 2021. Perhaps this is an indication as to how long it might still take :) About end of July.

**Comment by afariasfermin** [comment]
Any clue when this get deployed?

**Comment by shadi1999** [comment]
Hello all

I did update my eb instance to PHP 8.1 manually using predeploy script

this is my file under ROOT/.platform/hooks/prebuild/02_my_commands.sh:


```sh
#!/usr/bin/env bash
sudo /usr/local/composer self-update 
sudo yum -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
sudo yum -y install https://rpms.remirepo.net/enterprise/remi-release-7.rpm
sudo yum -y install yum-utils
sudo yum -y install rpl
sudo yum-config-manager --setopt=remi.priority=1 --save
# Reinstall ImageMagick from remi to prevent dependency conflicts
sudo yum -y remove ImageMagick
sudo yum -y install ImageMagick
export LC_ALL=en_US.UTF-8
export LANG=en_US.UTF-8
sudo yum-config-manager --disable 'remi-php*'
sudo yum-config-manager --enable remi-php81
rpl [remi-php81] "[remi-php81]" /etc/yum.repos.d/remi-php81.repo -e
sudo yum upgrade -y php php-*
sudo yum -y install php-zip libz-dev libzip-dev 
sudo yum -y install libzip-devel.x86_64
pecl install zlib zip 
sudo service httpd restart
```

it works fine with me and I hope that it could help you

**Comment by ninjacoder96** [comment]
it is also the same with ec2 amazon linux extra , there's no php 8.1 under the hood only php 8.0. any update on this?

**Comment by snabela** [comment]
Yes please add support for 8.1 to ec2 amazon linux extras, as well as to codebuild

**Comment by VaguelyOnline** [comment]
This is getting ridiculous now.

**Comment by dealer-solutions-gene** [comment]
Just saw this today, 8.1 added!

![image](https://user-images.githubusercontent.com/50188785/183514386-e028a2f7-4bd2-49ab-81c0-84d8cd72c45d.png)

(_ap-southeast-2_)

**Comment by newadventure079** [comment]
Is there an easy way to upgrade to php 8.1 or do we have to create a new environment?

**Comment by andyexeter** [comment]
For those looking to upgrade an existing environment, this can be done via the AWS CLI.

**WARNING: Don't run this on a production environment without testing! And don't blame me if it breaks something**

```sh
aws elasticbeanstalk update-environment \
--solution-stack-name "64bit Amazon Linux 2 v3.4.0 running PHP 8.1" \
--environment-id "x-xxxxxxxxxx" \
--region "xx-xxxx-x"
```

**Comment by ihor-sviziev** [comment]
Do anyone know if the php 8.1 packages became available for regular ec2 instances?

**Comment by 1tonyca** [comment]
https://docs.aws.amazon.com/elasticbeanstalk/latest/relnotes/release-2022-08-08-al2.html

**Comment by jordanade** [comment]
> https://docs.aws.amazon.com/elasticbeanstalk/latest/relnotes/release-2022-08-08-al2.html

Happy day! Looks like there's a typo tho:

> **New!** — Introduced new PHP 8.1 platform branch, running on version [8.0.18](https://www.php.net/releases/8_0_18.php). For more information see [PHP 8.1 Released!](https://www.php.net/releases/8.1/en.php)
> 
> on the PHP website.

This should say "running on version [8.1.8](https://www.php.net/releases/8_1_8.php)"

**Comment by KZeni** [comment]
> Do anyone know if the php 8.1 packages became available for regular ec2 instances?

@ihor-sviziev Do you mean the amzn2extra repo? I can say that having updated everything just now and checking `amazon-linux-extras` has `php8.0=latest enabled [ =stable ]` with php8.1 not listed (whereas php7.4 is listed there as available so one would imagine that's where 8.1 would show [right?] and it isn't available yet.)

Maybe I'm missing a step here, and I sure would love to know if there is an official way of doing so that I'm just missing (I know there are alternative options for getting PHP 8.1, but I'd like to stick to `amazon-linux-extras`, if I can.) It does appear it's not listed as available via this method yet (please correct me if I'm wrong.)

**Comment by jordanade** [comment]
Am I missing something, or is the `pecl` command missing?

**Comment by RahulDey12** [comment]
> Am I missing something, or is the `pecl` command missing?

You need to install `php-pear` first

**Comment by dejagersh** [comment]
> Am I missing something, or is the `pecl` command missing?

I ran into the same issue.

> You need to install php-pear first

Thanks @RahulDey12 . Works with `sudo yum install php-pear`

**Comment by swiffer** [comment]
@thejager - how are you installing it? for me it is not working with php 8.1

================================================================================
2022-08-09 15:02:07,694 P1670 [INFO] 	 Package         Arch        Version                Repository             Size
2022-08-09 15:02:07,694 P1670 [INFO] 	================================================================================
2022-08-09 15:02:07,694 P1670 [INFO] 	Installing:
2022-08-09 15:02:07,694 P1670 [INFO] 	 php-pear        noarch      1:1.10.12-9.amzn2      amzn2-core            359 k
2022-08-09 15:02:07,694 P1670 [INFO] 	Installing for dependencies:
2022-08-09 15:02:07,694 P1670 [INFO] 	 php-cli         aarch64     8.0.20-1.amzn2         amzn2extra-php8.0     5.0 M
2022-08-09 15:02:07,694 P1670 [INFO] 	 php-common      aarch64     8.0.20-1.amzn2         amzn2extra-php8.0     1.2 M
2022-08-09 15:02:07,694 P1670 [INFO] 	 php-process     aarch64     8.0.20-1.amzn2         amzn2extra-php8.0      84 k
2022-08-09 15:02:07,694 P1670 [INFO] 	 php-xml         aarch64     8.0.20-1.amzn2         amzn2extra-php8.0     175 k
2022-08-09 15:02:07,694 P1670 [INFO] 	
2022-08-09 15:02:07,694 P1670 [INFO] 	Transaction Summary
2022-08-09 15:02:07,694 P1670 [INFO] 	================================================================================
2022-08-09 15:02:07,694 P1670 [INFO] 	Install  1 Package (+4 Dependent packages)
2022-08-09 15:02:07,694 P1670 [INFO] 	
2022-08-09 15:02:07,694 P1670 [INFO] 	Total download size: 6.8 M
2022-08-09 15:02:07,694 P1670 [INFO] 	Installed size: 35 M

**Comment by KZeni** [comment]
@swiffer It appears you're encountering what I found & detailed in https://github.com/aws/elastic-beanstalk-roadmap/issues/214#issuecomment-1208871665 above.

That's showing you're using amzn2extra which then has PHP 8.0.20 being used, currently (that being the latest PHP version from the php8.0 item in `amazon-linux-extras` that's likely enabled & being used.)

I checked just now and `amazon-linux-extras` is still not listing php8.1 as an option. So it seems we're still waiting on AWS to complete that remaining item regarding PHP 8.1 being made available officially from them where past PHP versions are offered. Once that's offered (hopefully soon since it's now available elsewhere as mentioned by others above), one can go through the route of switching the enabled php version via `amazon-linux-extras` (among the other steps potentially involved.)

**Comment by RahulDey12** [comment]
> > Do anyone know if the php 8.1 packages became available for regular ec2 instances?
> 
> @ihor-sviziev Do you mean the amzn2extra repo? I can say that having updated everything just now and checking `amazon-linux-extras` has `php8.0=latest enabled [ =stable ]` with php8.1 not listed (whereas php7.4 is listed there as available so one would imagine that's where 8.1 would show [right?] and it isn't available yet.)
> 
> Maybe I'm missing a step here, and I sure would love to know if there is an official way of doing so that I'm just missing (I know there are alternative options for getting PHP 8.1, but I'd like to stick to `amazon-linux-extras`, if I can.) It does appear it's not listed as available via this method yet (please correct me if I'm wrong.)

They did not updated AL2 repository. But there is an Public AMI of elastic beanstalk with php81. Maybe that could help.

**Comment by PeterCai7** [comment]
> > https://docs.aws.amazon.com/elasticbeanstalk/latest/relnotes/release-2022-08-08-al2.html
> 
> Happy day! Looks like there's a typo tho:
> 
> > **New!** — Introduced new PHP 8.1 platform branch, running on version [8.0.18](https://www.php.net/releases/8_0_18.php). For more information see [PHP 8.1 Released!](https://www.php.net/releases/8.1/en.php)
> > on the PHP website.
> 
> This should say "running on version [8.1.8](https://www.php.net/releases/8_1_8.php)"

They have updated. Good call out.
![image](https://user-images.githubusercontent.com/39285101/183758263-8e5698f6-c2a0-4512-aafa-cf32dc894197.png)

**Comment by newadventure079** [comment]
> For those looking to upgrade an existing environment, this can be done via the AWS CLI.
> 
> **WARNING: Don't run this on a production environment without testing! And don't blame me if it breaks something**
> 
> ```shell
> aws elasticbeanstalk update-environment \
> --solution-stack-name "64bit Amazon Linux 2 v3.4.0 running PHP 8.1" \
> --environment-id "x-xxxxxxxxxx" \
> --region "xx-xxxx-x"
> ```

@andyexeter  If you run that command, you end up with this. It wants you to get on 3.4.0, even tho you're already on 3.4.0
![Screen Shot 2022-08-09 at 6 09 01 PM](https://user-images.githubusercontent.com/84156407/183782843-bcc2c009-b9b5-4d25-8974-07640eb50738.jpg)

**Comment by newadventure079** [comment]
> > For those looking to upgrade an existing environment, this can be done via the AWS CLI.
> > **WARNING: Don't run this on a production environment without testing! And don't blame me if it breaks something**
> > ```shell
> > aws elasticbeanstalk update-environment \
> > --solution-stack-name "64bit Amazon Linux 2 v3.4.0 running PHP 8.1" \
> > --environment-id "x-xxxxxxxxxx" \
> > --region "xx-xxxx-x"
> > ```
> 
> @andyexeter If you run that command, you end up with this. It wants you to get on 3.4.0, even tho you're already on 3.4.0 ![Screen Shot 2022-08-09 at 6 09 01 PM](https://user-images.githubusercontent.com/84156407/183782843-bcc2c009-b9b5-4d25-8974-07640eb50738.jpg)

It looks like that message goes away if you refresh the page or click out of the page and come back in

**Comment by RahulDey12** [comment]
Please confirm when it will available for the `amazon-linux-extras`

**Comment by iwishiwala** [comment]
They added PHP8.1 yesterday!

**Comment by KZeni** [comment]
> Please confirm when it will available for the `amazon-linux-extras`

Agreed. I'm looking forward to having an email notification via this issue from whoever finds out it's been added.

Here's the alphabetical list of what `amazon-linux-extras` currently offers on an otherwise up-to-date Amazon Linux 2 instance: https://gist.github.com/KZeni/9fa3510fb721e7782cdc3f89489fda8e
*Note that PHP has a dedicated item for each x.1 version that then auto-updates the x.x.1 version for patches where it's still missing php8.1 as of now.

**Comment by jordanade** [comment]
In addition to the missing `pecl`, which is easily remedied, it seems that PHP's `gd` and `pdo` extensions are also missing. Is this intentional, or is this upgrade half-baked? Should I workaround or wait?

**Comment by cloudjareen** [comment]
I would like to confirm that `gd` and `pdo` extensions do exist. Please ssh into the instance and check using the command `php -m`. Also thanks for the 8.1 release. However its very depressing to see such a highly voted feature take so much time to land. There are a lot of folks out here who do not use `docker` , coz otherwise they could have used any service besides beanstalk. But I would like everybody to invest learning docker, so that we are not at mercy of late releases like this. Thanks

**Comment by christian-manrique** [comment]
Hi all, thanks for all the updates but i am also having the same error, the instance type is `m6g.8xlarge` and since i updated all the package the `amazon-linux-extras list | grep php`  just return as latest 8.0 instead 8.1 which is what am looking forward to do, is there a workaround or suggestion to do? thanks for any advice

![Screen Shot 2022-08-11 at 8 12 16 PM](https://user-images.githubusercontent.com/992281/184262718-845c38da-50a3-4a5d-8a3f-4ce39702bc64.png)

**Comment by amman20** [comment]
@jordanade what is the workaround for the missing pcntl?

**Comment by MortenDHansen** [comment]
Alright. I i'm not able to grasp how to get extensions that are not pre-loaded to work. Many applications will for example need a redis connection (in my case elasticache), but there seems to be no way to install redis for php after the upgrade, no articles seems to be able to help. (eg. https://aws.amazon.com/premiumsupport/knowledge-center/elastic-beanstalk-pecl7-amazon-linux2/)

Running installation for any extension picks the 8.0 extension.

How can i uninstall the elasticbeanstalk php so i can reinstall from remi instead?

**Comment by andyexeter** [comment]
@MortenDHansen I ended up switching to [Pickle](https://github.com/FriendsOfPHP/pickle) after seeing pecl was no longer installed by default.

Here's prebuild script which sets the redis extension up (create it as e.g. `.platform/prebuild/10_install_redis.sh`):

```sh
#!/bin/bash

set -xe

if ! [ -f /bin/pickle ]; then
  wget https://github.com/FriendsOfPHP/pickle/releases/latest/download/pickle.phar -O /bin/pickle
  chmod +x /bin/pickle
fi

if ! php -r 'exit(extension_loaded("redis") ? 0 : 1);'; then
  /bin/pickle install redis --no-interaction --no-ansi
  echo 'extension=redis.so' > /etc/php.d/41-redis.ini
fi
```

**Comment by MortenDHansen** [comment]
> 

Oh thanks! - that mitigates an issue on the list. Can't seem to do pcntl that way though :)

**Comment by localpath** [comment]
You can install pecl then redis via it as well. Something to this effect

```bash
sudo yum install php-pear

sudo pecl channel-update pecl.php.net

sudo yum install php-devel

sudo no | sudo pecl install redis

sudo sed -i -e '/extension="redis.so"/d' /etc/php.ini

sudo echo 'extension="redis.so"' > /etc/php.d/41-redis.ini
```

**Comment by yurik94** [comment]
I have a few issues (tested on latest EB PHP 8.1 t4g graviton/arm and x64):

- pecl binary is not working (but i found `php /usr/lib/php/peclcmd.php` working instead)
- php 8.1 doesn't have pcntl support
- amazon extra linux php 8.1 repository is not present)

**Comment by publiux** [comment]
This is absolutely insane. I was planning on migrating to AWS services with EB, but this lack of support for basic PHP is just driving me to another provider at this point.

**Comment by nickdnk** [comment]
> https://docs.aws.amazon.com/elasticbeanstalk/latest/relnotes/release-2022-08-08-al2.html

@1tonyca While I'm sure we all appreciate that you now have 8.1 support, do you honestly feel like the communication with the community in this thread is adequate? No explanation for the absurd delay? No replies to/help with any of the problems people are having, the bugs found or missing features such as pecl, extra repo support etc?

**Comment by PeterCai7** [comment]
Sorry that we are not able to make the community happy at this time. We have noticed the bugs and missing features and thank you all for the feedbacks. But for extra repo support, this is the first time EB team was curating PHP runtime by our own rather than using Amazon Linux repo. We don't have a good answer for this point yet. We are working on all those stuff and explanary wordings. Will update once we feel they are verified and reliable.

**Comment by yurik94** [comment]
As REMI repository for php 8.1 is not working on arm/graviton cpus, I managed to install the missing php extensions (redis, ev, pcntl support) with the following yaml snippet of `.ebextensions`:
```
02_install_pecl_redis:
    command: |
      php /usr/lib/php/peclcmd.php install redis
      echo "extension=redis.so" | tee /etc/php.d/50-redis.ini
    test: "php -r \"exit(extension_loaded('redis') ? 1 : 0);\""

  03_install_pecl_ev:
    command: |
      php /usr/lib/php/peclcmd.php install ev
      echo "extension=ev.so" | tee /etc/php.d/50-ev.ini
    test: "php -r \"exit(extension_loaded('ev') ? 1 : 0);\""

  04_install_pcntl:
    command: |
      wget https://www.php.net/distributions/php-8.1.9.tar.gz
      tar -xf php-8.1.9.tar.gz
      cd php-8.1.9/ext/pcntl
      phpize
      ./configure
      make
      mv modules/pcntl.so /lib/php/extensions/no-debug-zts-20210902/pcntl.so
      cd ../../../
      rm -rf php-8.1.9
      echo "extension=pcntl.so" | tee /etc/php.d/50-pcntl.ini
    test: "php -r \"exit(extension_loaded('pcntl') ? 1 : 0);\""
```

Of course the hardocoded folder doesn't look good (you can use something like `php-config --extension-dir`) but it's tested and fully working on both x64 and graviton arm ATM

**Comment by christian-manrique** [comment]
Hi all, is this still not resolved? i want to deploy PHP 8.1on my ec2 server for the perks in performance and raised index of requests per second :'( please help

**Comment by PeterCai7** [comment]
https://docs.aws.amazon.com/elasticbeanstalk/latest/relnotes/release-2022-08-31-al2.html

**Comment by ihor-sviziev** [comment]
Could someone confirm, does php 8.1 became available through Amazon extra repositories on regular ec2 instances?

**Comment by salvoaranzulla** [comment]
@ihor-sviziev I don't see it:

$ amazon-linux-extras | grep "php8"
  _  php8.0                   available    [ =stable ]

**Comment by christian-manrique** [comment]
> @ihor-sviziev I don't see it:
> 
> $ amazon-linux-extras | grep "php8" _ php8.0 available [ =stable ]

did you execute the `sudo yum/apt-get update`?

**Comment by salvoaranzulla** [comment]
@christian-manrique Few minutes ago, I have done:

$ sudo yum update -y amazon-linux-extras

but nothing is changed:

$ amazon-linux-extras | grep "php8"
  _  php8.0                   available    [ =stable ]

**Comment by MaximGns** [comment]
Still the same here:

sudo amazon-linux-extras | grep "php8"
  _  php8.0                   available      [ =stable ]

**Comment by AlekMosingiewicz** [comment]
I second that. In addition, we get the following error:
`undefined symbol: executor_globals`
when trying to run your Elasticache-Client library for Linux. This blocks our attempts to upgrade our container to PHP 8.1, as Memcache extension is essential for our application.
When are you planning to release extras for 8.1?

**Comment by heathdutton** [comment]
Guys you can make a pecl alias using your ebextensions like so:
```
    "/usr/bin/pecl":
        mode: "000755"
        owner: root
        group: root
        content: |
            #!/bin/sh
            exec /usr/bin/php -C \
                -d disable_functions=none \
                -d include_path=/usr/lib/php \
                -d date.timezone=UTC \
                -d output_buffering=1 \
                -d variables_order=EGPCS \
                -d safe_mode=0 \
                -d register_argc_argv="On" \
                /usr/lib/php/peclcmd.php "$@"
```

**Comment by afariasfermin** [comment]
I ran into this issue for a couple of month and the best solution for me to enter in peaceful state of mind was learn about containers and migrate to a docker environment. 😄

**Comment by caseyw** [comment]
According to the [docs](https://docs.aws.amazon.com/elasticbeanstalk/latest/relnotes/release-2022-08-31-al2.html), version 2.0.20220805.0 contains PHP 8.1.9. When I pull the image and lookup the versions available:

`$ docker run -it --rm amazonlinux:2.0.20220805.0 amazon-linux-extras list | grep php`

Output:
 42  php7.4                   available    [ =stable ]
 51  php8.0                   available    [ =stable ]

Is this accurate based on the approach I'm taking, or am I going at this wrong?

Thanks

**Comment by peachesontour** [comment]
So this says PHP 8.1.11 is now available via Elastic Beanstalk:
https://docs.aws.amazon.com/elasticbeanstalk/latest/relnotes/release-2022-10-03-linux.html

But apparently not this way:
`amazon-linux-extras list | grep php`
Output:
  _  php7.4                   available    [ =stable ]
 51  php8.0=latest            enabled      [ =stable ]
 
 Is there an easy way to update? Or best to migrate to AL 2022 or Ubuntu?

**Comment by andres-montanez** [comment]
Apparently pulling from this image `amazonlinux:2022` does the trick.
No need to use the extras, either.
```
bash-5.1# yum install -y php-cli
Last metadata expiration check: 0:02:09 ago on Sat Oct  8 15:51:09 2022.
Dependencies resolved.
=============================================================================================================================================================================================================================================================================================================
 Package                                                                Architecture                                                       Version                                                                             Repository                                                               Size
=============================================================================================================================================================================================================================================================================================================
Installing:
 php8.1-cli                                                             x86_64                                                             8.1.7-1.amzn2022.0.1                                                                amazonlinux                                                             3.5 M

Transaction Summary
=============================================================================================================================================================================================================================================================================================================
Install  1 Package

Total download size: 3.5 M
Installed size: 16 M
Downloading Packages:
php8.1-cli-8.1.7-1.amzn2022.0.1.x86_64.rpm                                                                                                                                                                                                                                   1.1 MB/s | 3.5 MB     00:03
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                                                                                                                                                        832 kB/s | 3.5 MB     00:04
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                                                                                                                                                                     1/1
  Installing       : php8.1-cli-8.1.7-1.amzn2022.0.1.x86_64                                                                                                                                                                                                                                              1/1
  Running scriptlet: php8.1-cli-8.1.7-1.amzn2022.0.1.x86_64                                                                                                                                                                                                                                              1/1
  Verifying        : php8.1-cli-8.1.7-1.amzn2022.0.1.x86_64                                                                                                                                                                                                                                              1/1

Installed:
  php8.1-cli-8.1.7-1.amzn2022.0.1.x86_64

Complete!
bash-5.1# php -v
PHP 8.1.7 (cli) (built: Jun  7 2022 18:21:38) (NTS gcc x86_64)
Copyright (c) The PHP Group
Zend Engine v4.1.7, Copyright (c) Zend Technologies
```
🤘

**Comment by RahulDey12** [comment]
> So this says PHP 8.1.11 is now available via Elastic Beanstalk: https://docs.aws.amazon.com/elasticbeanstalk/latest/relnotes/release-2022-10-03-linux.html
> 
> But apparently not this way: `amazon-linux-extras list | grep php` Output: _ php7.4 available [ =stable ] 51 php8.0=latest enabled [ =stable ]
> 
> Is there an easy way to update? Or best to migrate to AL 2022 or Ubuntu?

use this syntax

```bash
rpm -i /opt/elasticbeanstalk/RPMS/php-debuginfo-8.1.8-1.amzn2.x86_64.rpm
```

**Referance:**
https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/create_deploy_PHP.container.html#php-8-1-considerations

**Comment by KZeni** [comment]
@andres-montanez Okay, I see that https://aws.amazon.com/amazon-linux-2/faqs/ points to Amazon Linux 2022 (AL2022) as being the successor to Amazon Linux 2 (AL2), but AL2022 is still showing as Release Candidate per https://aws.amazon.com/linux/amazon-linux-2022/.

Also, the AMI Catalog doesn't really feature it with the only match for "Amazon Linux 2022" or "AL2022" either being the one "Amazon ECS-Optimized Amazon Linux 2022 (AL2022) x86_64 AMI - PREVIEW" AMI under "AWS Marketplace AMIs" or a handful of AMIs being under the "Community AMIs" tab (it then not being in the recommended/"Quickstart AMIs" at all [likely due to it not being officially released yet.])

I think I'd prefer to stay on Amazon Linux 2 for the meantime (with the new EOL being June 30th, 2024 per their FAQs linked above) with this just needing PHP 8.1 added to Amazon Linux Extras rather than jumping over to a wholly different OS version that's not officially released yet (at least not considered ready for production environments per it still being a release candidate.) Especially considering https://docs.aws.amazon.com/linux/al2022/ug/compare-al2-to-AL2022.html shows migrating between the two not being the most seamless experience in certain ways (at least for now.)

@RahulDey12 Doesn't that method just avoid the use of Amazon Linux Extras? Ideally, I'd like to keep using the Amazon Linux Extras convention that was established for Amazon Linux 2 since we just need them to add support for new PHP versions via this tool when they're made available (especially after elastic beanstalk has it added).

It would seem odd for them to drop support for adding new versions of things to amazon-linux-extras considering that's the convention Amazon Linux 2 set up for these items with Amazon Linux 2 then saying it's still being supported through mid-2024.

**As far as I can tell**, we're still waiting on PHP 8.1.x to be added to Amazon Linux Extras. We technically have yet to be given a confirmation that AWS has any plans to do so (let alone provide a timeline.) They got it added to EB (~9 months after release) while mentions of amazon-linux-extras have either been effectively ignored, had existing customers say they also are waiting for it, or had people bring up ways to add PHP 8.1 that essentially avoids the use of amazon-linux-extras to do so. It's also been brought up on re:Post multiple times (also without luck when it comes to them saying they'll actually be adding PHP 8.1 to Amazon Linux 2 or not [both in providing a timeline & just saying they're planning to do it at all.])
1. https://repost.aws/questions/QUsfhDbww4SAy8i5Jmd4vcMg/amazon-linux-2-amazon-linux-extras-php-8-1-support
2. https://repost.aws/questions/QUYOYerObQQ36CDxBsry62Ow/how-to-request-php-8-1-to-be-added-to-amazon-linux-extras
3. https://repost.aws/questions/QUEE13bbWpRICK4skmyN706g/when-is-php-8-1-coming-to-amazon-linux-extras

**Comment by RahulDey12** [comment]
@KZeni I think the `amazon-linux-extra` is a wonderful way to install any version of PHP. But from PHP 8.1 I think AWS is not planning to add any future support for PHP in the amazon-linux-extras repo. So I think it's time to move on RHEL distro instead of AL2. AL2 was also an RHEL-based distro so there is not much of a change in terms of configuration. But the draw back is that RHEL distros cost more money than AL2 because of the licensing. But RHEL has wonderful support for PHP thanks to Remi Repos.

**Comment by KZeni** [comment]
@RahulDey12 Have they actually stated they're no longer supporting `amazon-linux-extras` or are you just guessing that's the case?

Also, when it comes to changing over to a different Linux distro... it's kinda like seeing people asking for an update be made to a key feature of an operating system that AWS has stated they're supporting through mid-2024 (like the case is here) and then just saying to switch operating systems as a resolution (while then only stating a specific OS/distro when there's actually a successor to the OS/distro in question [AL2 => AL2022] that would also address the issue.) Seems a bit heavy handed of a recommendation for a case like this (especially when there are other options that don't have added licensing fees, are also reasonable & reputable, and wouldn't have this issue [ex. Ubuntu, Fedora, CentOS, etc. and even the successor to Amazon Linux 2... Amazon Linux 2022 which is then based on Fedora which is then closely related to RHEL where it just needs to leave Release Candidate/Preview status.])

That said, the fact AWS (both EB & the AL2 extras) hasn't managed this very well has me looking outside of their offerings if they keep lagging behind like this (ex. Docker, using a different linux distro, etc.... possibly keep with the latest Amazon Linux distros [or possibly use Amazon ECS] while then just using Docker for the server software/environment.)

**Comment by faizanakram99** [comment]
**EDIT: nvm, works on fresh environment ... it was something specific to an environment which was upgraded via a command** 

![image](https://user-images.githubusercontent.com/10880992/196743535-6f3cc113-92c6-4c70-b279-ed45d0d35015.png)

~~Does this really work for anyone ?~~

~~when i do php -v, it shows php 8 version instead of 8.1~~

![image](https://user-images.githubusercontent.com/10880992/196743679-e3a331b0-c262-470a-959d-2b34822c6756.png)

~~What is wrong ?~~

~~@RahulDey12 i don't see that directory, there is no RPMS directory inside /opt/elasticbeanstalk~~

**Comment by peachesontour** [comment]
It finally appeared today:
`amazon-linux-extras list | grep php`
`  _  php7.4                   available    [ =stable ]`
` 51  php8.0=latest            enabled      [ =stable ]`
`  _  php8.1                   available    [ =stable ]`

**Comment by christian-manrique** [comment]
i am able to upgrade my ec2 instance to php 8.1, thanks for the Support team very nice and appreciate it
```
$ amazon-linux-extras list | grep php
  _  php7.4                   available    [ =stable ]
 43  php8.0=latest            enabled      [ =stable ]
  _  php8.1                   available    [ =stable ]
```

**Comment by tobiaseichert** [comment]
Running the default `/bin/sh -c composer.phar install --no-ansi --no-interaction` inside a PHP 8.1 environment (64bit Amazon Linux 2/3.5.0) currently results in an error, e.g.

```
In Git.php line 484:
                                                                               
  Failed to clone https://git.drupalcode.org/project/coder.git, git was not f  
  ound, check that it is installed and in your PATH env.                       
                                                                               
  sh: git: command not found 
```

However, git is installed:
```
[ec2-user@ip-172-31-4-122 ~]$ which git
/usr/bin/git
```
 Is this an environment related bug at deployment time? There's no problem when performing a deployment inside a PHP 7.4 environment.

**Comment by RahulDey12** [comment]
New issue is coming during update

```bash
Could not retrieve mirrorlist https://amazonlinux-2-repos-ap-south-1.s3.dualstack.ap-south-1.amazonaws.com/2/extras/php8.1/latest/x86_64/mirror.list error was
14: HTTPS Error 403 - Forbidden
```

Looks like they have blocked access to the rpms

**Comment by christian-manrique** [comment]
> New issue is coming during update
> 
> ```shell
> Could not retrieve mirrorlist https://amazonlinux-2-repos-ap-south-1.s3.dualstack.ap-south-1.amazonaws.com/2/extras/php8.1/latest/x86_64/mirror.list error was
> 14: HTTPS Error 403 - Forbidden
> ```
> 
> Looks like they have blocked access to the rpms

Despite my ec2 instance is pulling php8.1 as available through amazon-linux-extras  i am having the same issue
```Could not retrieve mirrorlist https://amazonlinux-2-repos-ap-south-1.s3.dualstack.ap-south-1.amazonaws.com/2/extras/php8.1/latest/x86_64/mirror.list error was 14: HTTPS Error 403 - Forbidden```
, the CPU Architecture of my ec2 instance is  aarch64 (64x) and  only the support provided by amazon-linux-extras remis is for x86_64 CPU, seems theres not any builder for aarch64 CPUs this link provide more information https://forum.remirepo.net/viewtopic.php?id=4016, the remis repositories for php8.1 only has available x86_64/ folder https://fr2.rpmfind.net/linux/remi/enterprise/7/php81/, any advice is very pleased 

![Screen Shot 2022-10-27 at 10 34 13 AM](https://user-images.githubusercontent.com/992281/198317403-fe8ed81c-8b9a-4cab-9e8b-167fffa3c962.png)

**Comment by RahulDey12** [comment]
> > New issue is coming during update
> > ```shell
> > Could not retrieve mirrorlist https://amazonlinux-2-repos-ap-south-1.s3.dualstack.ap-south-1.amazonaws.com/2/extras/php8.1/latest/x86_64/mirror.list error was
> > 14: HTTPS Error 403 - Forbidden
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Looks like they have blocked access to the rpms
> 
> Despite my ec2 instance is pulling php8.1 as available through amazon-linux-extras i am having the same issue `Could not retrieve mirrorlist https://amazonlinux-2-repos-ap-south-1.s3.dualstack.ap-south-1.amazonaws.com/2/extras/php8.1/latest/x86_64/mirror.list error was 14: HTTPS Error 403 - Forbidden` , the CPU Architecture of my ec2 instance is aarch64 (64x) and only the support provided by amazon-linux-extras remis is for x86_64 CPU, seems theres not any builder for aarch64 CPUs this link provide more information https://forum.remirepo.net/viewtopic.php?id=4016, the remis repositories for php8.1 only has available x86_64/ folder https://fr2.rpmfind.net/linux/remi/enterprise/7/php81/, any advice is very pleased
> 
> ![Screen Shot 2022-10-27 at 10 34 13 AM](https://user-images.githubusercontent.com/992281/198317403-fe8ed81c-8b9a-4cab-9e8b-167fffa3c962.png)

Remi doesn't provide any build for ARM CPU.

**Comment by laurent-le-graverend** [comment]
> It finally appeared today: `amazon-linux-extras list | grep php`

On what AMI do you see it @peachesontour ?

I don't see it yet on the latest Arm64 AMI (Amazon Linux 2 LTS Arm64 Kernel 5.10 AMI 2.0.20221004.0 arm64 HVM gp2
ami-081dc0707789c2daf)
<img width="497" alt="image" src="https://user-images.githubusercontent.com/332246/198814691-9c284bdd-e4c3-4df3-866a-1764c4876cb5.png">

**Comment by peachesontour** [comment]
> 

Amazon Linux 2 AMI 2.0.20210427.0 x86_64 HVM gp2

But I get the `Could not retrieve mirrorlist` error mentioned above while trying to update too. Does anyone have a solution?

**Comment by RahulDey12** [comment]
This is now ridiculous guys, my production deployment is stuck for 2 months. I and my team member's patients are breaking now. What a terrible experience. I think this is the time when I am gonna switch to some other distro.

**Comment by KZeni** [comment]
@laurent-le-graverend:
> > It finally appeared today: `amazon-linux-extras list | grep php`
>
> On what AMI do you see it @peachesontour ?
>
> I don't see it yet on the latest Arm64 AMI (Amazon Linux 2 LTS Arm64 Kernel 5.10 AMI 2.0.20221004.0 arm64 HVM gp2
ami-081dc0707789c2daf)
> ![](https://user-images.githubusercontent.com/332246/198814691-9c284bdd-e4c3-4df3-866a-1764c4876cb5.png)

This is weird.

I'm seeing the same thing (PHP 8.1 isn't listed via `amazon-linux-extras`.) Oddly, I saw PHP 8.1 being offered via `amazon-linux-extras` on this same EC2 instance a week ago or so back when people were reporting it was made available.

So... **did AWS remove PHP 8.1 from `amazon-linux-extras` just a few days after adding it?** Or why else is it no longer being shown when this same instance was showing it as offered last week (didn't make the transition during the few days it was available [and kinda glad I didn't if they removed that offering for a good reason])? Is it planning to be made available again?

**Comment by Ben-CA** [comment]
Looking at upgrading from PHP 7.4 to PHP 8.1 on AWS EB - is there a way to upgrade this yet (preferably from the console) without needing to setup an entirely new environment? Like, surely by now there is some sort of a quick "migration upgrade" path?
![image](https://user-images.githubusercontent.com/47060874/199119414-38280ef3-2724-45fa-a8f7-e6a8ca7b5ad2.png)

![image](https://user-images.githubusercontent.com/47060874/199119142-d1f1297d-af55-42d1-9b72-757dc3213fc0.png)

Why can't I upgrade to one of these with a few clicks?

![image](https://user-images.githubusercontent.com/47060874/199120261-49e2c789-c67e-4255-a4a1-57ec28462e9d.png)

**Comment by JvH69** [comment]
> Looking at upgrading from PHP 7.4 to PHP 8.1 on AWS EB - is there a way to upgrade this yet (preferably from the console) without needing to setup an entirely new environment? Like, surely by now there is some sort of a quick "migration upgrade" path?

The solution as described by andyexeter worked for me

**Comment by faizanakram99** [comment]
@Ben-CA it is risky to do inplace upgrade, create a new environment and then swap cname ... see blue green deployment docs of aws beanstalk

**Comment by Ben-CA** [comment]
I know there's the command line option that some people have had success with, and in the past I've created a new environment - but my environment has a lot of environment variables, and to my knowledge these all need to be manually copied and pasted to the new one. At least last time I did it (months ago) while you could save the configuration, it won't let you pick the updated platform version if you try to use the saved configuration - so that's not useable.

I find it bizarre that AWS doesn't have a better way of doing this - why isn't there a "change environment platform" option in the console, that could automate creating a blue/green deployment with all the same settings?

And @faizanakram99 - my apps only use a single instance - what are the risks of an in-place platform update? (I would typically do this in off-hours anyways and test afterwards; and worst case I just need to re-create the environment?) I have a number of separate applications to do, and there's just so much room for making a mistake with manually setting up a new environment, copying all the variables, etc.

**Comment by Ben-CA** [comment]
I just tested the upgrade on one of my demo environments (going from PHP 7.4) via the command line update.

Opened AWS CloudShell, and pasted in:
**aws elasticbeanstalk update-environment --solution-stack-name "64bit Amazon Linux 2 v3.5.0 running PHP 8.1" --environment-id "e-XXXXXXX" --region "XX-XXXXXX-1"**

Pressed enter, and it took maybe 2 minutes max - went grey in the console, then came back healthy. Had a total of perhaps 30 seconds of downtime for my application (demo). It automatically created the new instance, and once healthy, removed the old instance.

What would be the downside of using this on production environments on a weekend when nobody will notice the 30 seconds of downtime? (I've spent hours to accomplish the same thing "manually" via creating a new environment in the console.)

**Comment by faizanakram99** [comment]
@Ben-CA make sure the installed version of PHP in that demo environment is 8.1 after upgrade.

The in place upgrade command isn't always smooth,  in one of our apps the installed version was stuck on PHP 8 despite beanstalk showing it as PHP 8.1 version.

I agree with you though, copying environment variables is a painful from console UI, might be a good idea to invest in IaC like pulumi, or just plain bash scripts using aws cli where you can supply env vars via a file

**Comment by Ben-CA** [comment]
Thanks @faizanakram99; I checked via PHP info() and it says it's PHP 8.1.11 now, so seems to have worked.

![image](https://user-images.githubusercontent.com/47060874/199374324-9d4f04d4-4757-4a5d-863e-7999d50d00f8.png)

**Comment by Ben-CA** [comment]
Just updating this for future reference; that command worked great for 3 environments where I was going from PHP 7.4, but only "pretended" to update on the 4th one that I tried upgrading from 8.0. Like @faizanakram99 encountered, it shows in the AWS console that it's on 8.1 now:

![image](https://user-images.githubusercontent.com/47060874/201428225-0c977a37-4801-4d44-a302-9715be741b6d.png)

While a php info() running on the server shows it hasn't actually been updated:

![image](https://user-images.githubusercontent.com/47060874/201428383-5d12863d-9c5e-4a66-9701-e228d26ddbac.png)

**Comment by andryska** [comment]
@Ben-CA we had the same issue and realized that there was an Image ID hardcoded on the environment:
```
AWSEBAutoScalingLaunchConfiguration.aws:autoscaling:launchconfiguration:
  ImageId: ami-00cf32aa5968ac2fd
```

This ImageId contains PHP 8.0.8 so it would never update even if ElasticBeanstalk thinks it's updating.

Save your config locally with `eb config save` and see if you have the same.

**Comment by Ben-CA** [comment]
@andryska  - I noticed there was an AMI set for it, and wondered if perhaps that was the issue. (And haven't got around to investigating further yet) Were you able to remove that AMI from the environment config and then it worked?

**Comment by andryska** [comment]
> @andryska - I noticed there was an AMI set for it, and wondered if perhaps that was the issue. (And haven't got around to investigating further yet) Were you able to remove that AMI from the environment config and then it worked?

Yes. Just remove the two lines and then `eb create env-name --cfg saved-config-name` and it will create it with the latest PHP version.

**Comment by Ben-CA** [comment]
Thanks; I'll have to give that a try. (I was thinking of trying from the console actually; that's where I noticed the AMI set.)

**Comment by KZeni** [comment]
**For those using Amazon Linux 2 with `amazon-linux-extras`, it looks like PHP 8.1 has reappeared!**

![2022-12-12](https://user-images.githubusercontent.com/118200/207137880-2ee31be8-031d-4cfb-8a30-d3c241a6583e.png)

It's also been there for a few days now (unlike the last time it was available where it was removed a couple of days later... though, I imagine it could still disappear again while I have my doubts they'd do that a second time.)

So it seems like we now have PHP 8.1.x via `amazon-linux-extras` just in time for PHP 8.0.x to be entering security-only support and PHP 8.2.x having been released. Kinda wondering if they're going to keep this recent update going and also loop PHP 8.2.x into this to be available as well at some point soon, but I wouldn't be surprised if they make this a trend of only releasing the next version of PHP once the newest version is available & the PHP version they're offering via `amazon-linux-extras` (ex. we'll get 8.2.x once 8.1.x enters security-only support & 8.3.x is released in about a year from now... while I can't say one way or the other, the timing of this becoming available on this particular date seems like more than a coincidence.)

Meanwhile, **Amazon Linux 2 appears to have extended its EOL date yet again so it's now June 30th, 2025** per https://aws.amazon.com/amazon-linux-2/faqs/ (it seems they missed a few dates still saying 2024 in some spots of the FAQ while the main date now does say 2025) so we'll likely have 2 more major PHP version releases needing to be handled via `amazon-linux-extras` yet (per them stating they'll still be supporting that OS & its core components [as the FAQ states as well] until then.) This is probably due to Amazon Linux 2022 (AL2022) which is the successor to AL2 still being in release candidate status (per https://aws.amazon.com/linux/amazon-linux-2022/) while then probably needing to allow for a reasonable amount of time to transition from one to the other without one being unsupported and/or one not being deemed ready for production use.

**Comment by lutzklai** [comment]
@Ben-CA To accomplish the upgrade from 8.0 to 8.1 I actually just used the AWS GUI.

Steps:
1. In the beanstalk service, save beanstalk configuration
2. goto your beanstalk s3 bucket to download the template
3. in a text editor, manually update the platform version in the template
4. upload the template back to s3
5. In the beanstalk service, load the updated template back to the same environment

and it just spins down the existing EC2 instances that contains php 8.0 and spins up an 8.1 instance. That config actually copies everything including environment variables so you could do the whole blue-green deployment under a new environment if you wanted to. But the in-place upgrade worked for my use case.

I faced a similar issue to you and @andryska where every time I deployed it reverted the instance back to the php 8.0 platform. It turned out I had the `SolutionStack` variable in my `env.yaml` file pointing to a php 8.0 stack instead of 8.1. Updating that variable to 8.1 fixed it.

**Comment by Ben-CA** [comment]
Thanks @lutzklai - I used the AWS GUI as well to change AMI (SolutionStack) that was specified in the config to an AMI that had PHP 8.1 instead. Different method than you describe, but likely accomplished essentially the same thing in the backend. (I'm not sure why this AMI is statically assigned, but I didn't find a simple way to avoid that.)
